<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Runs a principal components analysis, the facile way. — fpca • FacileAnalysis</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Runs a principal components analysis, the facile way. — fpca" />
<meta property="og:description" content="Performs a principal components analysis over a specified assay from the
(subset of) samples in a FacileDataStore." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FacileAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/FacileAnalysis.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/FacileAnalysis-RNAseq.html">RNA-seq Analysis: The Facile Way</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-examples.html">Reproducing genomic anlayses the facile way</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fdge.html">Facile Differential Expression Analysis</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fpca.html">Facile Principal Components Analysis</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fsea.html">Facile (Gene) Set Expression Analysis</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Runs a principal components analysis, the facile way.</h1>
    
    <div class="hidden name"><code>fpca.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs a principal components analysis over a specified assay from the
(subset of) samples in a FacileDataStore.</p>
    </div>

    <pre class="usage"><span class='fu'>fpca</span>(<span class='no'>x</span>, <span class='kw'>assay_name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>dims</span> <span class='kw'>=</span> <span class='fl'>5</span>, <span class='kw'>ntop</span> <span class='kw'>=</span> <span class='fl'>500</span>,
  <span class='kw'>row_covariates</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>col_covariates</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>batch</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>main</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)

<span class='co'># S3 method for facile_frame</span>
<span class='fu'>fpca</span>(<span class='no'>x</span>, <span class='kw'>assay_name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>dims</span> <span class='kw'>=</span> <span class='fl'>5</span>,
  <span class='kw'>ntop</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>row_covariates</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>col_covariates</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>batch</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>main</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>custom_key</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Sys.getenv.html'>Sys.getenv</a></span>(<span class='st'>"USER"</span>), <span class='no'>...</span>)

<span class='co'># S3 method for matrix</span>
<span class='fu'>fpca</span>(<span class='no'>x</span>, <span class='kw'>dims</span> <span class='kw'>=</span> <span class='fl'>5</span>, <span class='kw'>ntop</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>row_covariates</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>col_covariates</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>batch</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>main</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>use_irlba</span> <span class='kw'>=</span> <span class='no'>dims</span> <span class='kw'>&lt;</span> <span class='fl'>7</span>, <span class='kw'>center</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>scale.</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a facile data container (FacileDataSet), or a <code>facile_frame</code>
(refer to the FacileDataStore (facile_frame) section.</p></td>
    </tr>
    <tr>
      <th>assay_name</th>
      <td><p>the name of the assay to extract data from to perform the
PCA. If not specified, default assays are taken for each type of assay
container (ie. <code>default_assay(facile container)</code>, <code>"counts"</code> for a
<code>DGEList</code>, <code>assayNames(SummarizedExperiment)[1L]</code>, etc.)</p></td>
    </tr>
    <tr>
      <th>dims</th>
      <td><p>the number of PC's to calculate (minimum is 3).</p></td>
    </tr>
    <tr>
      <th>ntop</th>
      <td><p>the number of features (genes) to include in the PCA. Genes are
ranked by decreasing variance across the samples in <code>x</code>.</p></td>
    </tr>
    <tr>
      <th>row_covariates, col_covariates</th>
      <td><p>data.frames that provie meta information
for the features (rows) and samples (columns). The default is to get
these values from "the obvious places" given <code>x</code> (<code>$genes</code> and <code>$samples</code>
for a DGEList, or the sample and feature-level covariate database tables
from a FacileDataSet, for example).</p></td>
    </tr>
    <tr>
      <th>batch, main</th>
      <td><p>specify the covariates to use for batch effect removal.
Refer to the <code><a href='https://rdrr.io/pkg/FacileData/man/remove_batch_effect.html'>FacileData::remove_batch_effect()</a></code> help for more information.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>an fpca result</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>FacilePcaAnalysisResult</code> produced here can be used in "the usual" ways,
ie. can be <code>viz</code>-ualized. <code><a href='FacileAnalysisResultViz.html'>shine()</a></code> is 1/4th-implemented, and <code><a href='FacileAnalysisResultViz.html'>report()</a></code>
has not been worked on yet.</p>
<p>Importantly / interestingly, you can shoot this result into <code><a href='ffsea.html'>ffsea()</a></code> to
perform gene set enrichment analysis over a specified dimension to identify
functional categories loaded onto differend PCs.</p>
    <h2 class="hasAnchor" id="development-notes"><a class="anchor" href="#development-notes"></a>Development Notes</h2>

    

<p>Follow progress on implementation of <code><a href='FacileAnalysisResultViz.html'>shine()</a></code> and <code><a href='FacileAnalysisResultViz.html'>report()</a></code> below:</p><ol>
<li><p><a href='https://github.com/facileverse/FacileAnalysis/issues/11'>Implement shine()</a></p></li>
<li><p><a href='https://github.com/facileverse/FacileAnalysis/issues/12'>Implement report()</a></p></li>
</ol>

<p>Note that there are methods defined for other assay containers, like an
<code><a href='https://rdrr.io/pkg/edgeR/man/DGEList.html'>edgeR::DGEList</a></code>, <code>limma::EList</code>, and <code>SummarizedExperiment</code>. If these are
called directly, their downstream use within the facile ecosystem isn't
yet fully supported. Development of the
<a href='https://github.com/facileverse/FacileBioc'>FacileBioc package</a>
will address this.</p>
    <h2 class="hasAnchor" id="random-things-to-elaborate-on"><a class="anchor" href="#random-things-to-elaborate-on"></a>Random Things to elaborate on</h2>

    

<p>The code here is largely inspired by DESeq2's plotPCA.</p>
<p>You should look at factominer:</p><ul>
<li><p>http://factominer.free.fr/factomethods/index.html</p></li>
<li><p>http://factominer.free.fr/graphs/factoshiny.html</p></li>
</ul>

    <h2 class="hasAnchor" id="teaching-and-tutorials"><a class="anchor" href="#teaching-and-tutorials"></a>Teaching and Tutorials</h2>

    


<p>This looks like a useful tutorial to use when explaining the utility of
PCA analysis:
http://alexhwilliams.info/itsneuronalblog/2016/03/27/pca/</p>
    <h2 class="hasAnchor" id="faciledatastore-facile-frame-"><a class="anchor" href="#faciledatastore-facile-frame-"></a>FacileDataStore (facile_frame)</h2>

    

<p>We enable the user to supply extra sample covariates that are not found
in the FacileDataStore associated with these samples <code>x</code> by adding them as
extra columns to <code>x</code>.</p>
<p>If manually provioded col_covariates have the same name as internal sample
covariates, then the manually provided ones will supersede the internals.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>efds</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>FacileData</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/FacileData/man/exampleFacileDataSet.html'>exampleFacileDataSet</a></span>()

<span class='co'># A subset of samples ------------------------------------------------------</span>
<span class='no'>pca.crc</span> <span class='kw'>&lt;-</span> <span class='no'>efds</span> <span class='kw'>%&gt;%</span>
  <span class='kw pkg'>FacileData</span><span class='kw ns'>::</span><span class='fu'>filter_samples</span>(<span class='no'>indication</span> <span class='kw'>==</span> <span class='st'>"CRC"</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'>fpca</span>()
<span class='kw'>if</span> (<span class='fu'><a href='https://rdrr.io/r/base/interactive.html'>interactive</a></span>()) {
  <span class='co'># report(pca.crc, color_aes = "sample_type")</span>
  <span class='fu'><a href='FacileAnalysisResultViz.html'>shine</a></span>(<span class='no'>pca.crc</span>)
  <span class='fu'><a href='FacileAnalysisResultViz.html'>viz</a></span>(<span class='no'>pca.crc</span>, <span class='kw'>color_aes</span> <span class='kw'>=</span> <span class='st'>"sex"</span>)
}

<span class='co'># Same PCA as above, but regress "sex" out of samples first.</span>
<span class='no'>pca.crcs</span> <span class='kw'>&lt;-</span> <span class='fu'>samples</span>(<span class='no'>pca.crc</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'>fpca</span>(<span class='kw'>batch</span> <span class='kw'>=</span> <span class='st'>"sex"</span>)
<span class='kw'>if</span> (<span class='fu'><a href='https://rdrr.io/r/base/interactive.html'>interactive</a></span>()) {
  <span class='fu'><a href='FacileAnalysisResultViz.html'>viz</a></span>(<span class='no'>pca.crcs</span>, <span class='kw'>color_aes</span> <span class='kw'>=</span> <span class='st'>"sex"</span>)
}

<span class='no'>pca.gdb</span> <span class='kw'>&lt;-</span> <span class='no'>pca.crc</span> <span class='kw'>%&gt;%</span>
  <span class='fu'><a href='signature.html'>signature</a></span>(<span class='kw'>dims</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fl'>3</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'>result</span>() <span class='kw'>%&gt;%</span>
  <span class='kw pkg'>multiGSEA</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/multiGSEA/man/GeneSetDb-class.html'>GeneSetDb</a></span>()

<span class='co'># All samples --------------------------------------------------------------</span>
<span class='no'>pca.all</span> <span class='kw'>&lt;-</span> <span class='fu'>fpca</span>(<span class='no'>efds</span>)
<span class='kw'>if</span> (<span class='fu'><a href='https://rdrr.io/r/base/interactive.html'>interactive</a></span>()) {
  <span class='fu'><a href='FacileAnalysisResultViz.html'>viz</a></span>(<span class='no'>pca.all</span>, <span class='kw'>color_aes</span> <span class='kw'>=</span> <span class='st'>"indication"</span>, <span class='kw'>shape_aes</span> <span class='kw'>=</span> <span class='st'>"sample_type"</span>)
  <span class='co'># report(pca.all, color_aes = "indication", shape_aes = "sample_type")</span>
}


<span class='co'># This works on "normal" DGELists, too. -----------------------------------</span>
<span class='no'>pca.dgelist</span> <span class='kw'>&lt;-</span> <span class='no'>efds</span> <span class='kw'>%&gt;%</span>
  <span class='fu'>filter_samples</span>(<span class='no'>indication</span> <span class='kw'>==</span> <span class='st'>"CRC"</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'>as.DGEList</span>() <span class='kw'>%&gt;%</span>
  <span class='fu'>fpca</span>()</div><div class='output co'>#&gt; <span class='error'>Error in filter_samples(., indication == "CRC"): could not find function "filter_samples"</span></div><div class='input'><span class='kw'>if</span> (<span class='fu'><a href='https://rdrr.io/r/base/interactive.html'>interactive</a></span>()) {
  <span class='fu'><a href='FacileAnalysisResultViz.html'>report</a></span>(<span class='no'>pca.dgelist</span>, <span class='kw'>color_aes</span> <span class='kw'>=</span> <span class='st'>"sample_type"</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#development-notes">Development Notes</a></li>
      <li><a href="#random-things-to-elaborate-on">Random Things to elaborate on</a></li>
      <li><a href="#teaching-and-tutorials">Teaching and Tutorials</a></li>
      <li><a href="#faciledatastore-facile-frame-">FacileDataStore (facile_frame)</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Steve Lianoglou, Denali Therapeutics.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


